simulation/amaru-sim/src/simulator/simulate.rs

```diff
@@ -242,69 +250,105 @@ impl<Msg> Drop for World<Msg> {
 }
 
 pub fn simulate<Msg, F>(
-    config: Config,
-    number_of_nodes: u8,
+    config: SimulateConfig,
     spawn: F,
-    generate_messages: impl Strategy<Value = Vec<Msg>>,
-    property: impl Fn(Trace<Msg>) -> Result<(), String>,
+    generator: impl Fn(&mut StdRng) -> Vec<Reverse<Entry<Msg>>>,
+    property: impl Fn(History<Msg>) -> Result<(), String>,
     trace_buffer: Arc<parking_lot::Mutex<TraceBuffer>>,
     persist_on_success: bool,
 ) where
     Msg: Debug + PartialEq + Clone + Serialize,
     F: Fn() -> NodeHandle<Msg>,
 {
-    let mut runner = TestRunner::new(config);
-    let now = Instant::at_offset(Duration::from_secs(0));
-    let generate_messages = (any::<u64>(), generate_messages).prop_map(|(seed, msgs)| {
-        let mut rng = StdRng::seed_from_u64(seed);
-        let arrival_times = generate_arrival_times(&mut rng, now, 200.0, msgs.len());
-        msgs.into_iter()
-            .enumerate()
-            .map(|(idx, msg)| {
-                Reverse(Entry {
-                    arrival_time: arrival_times[idx],
-                    envelope: Envelope {
-                        src: "c1".to_string(),
-                        dest: "n1".to_string(),
-                        body: msg,
-                    },
-                })
-            })
-            .collect()
-    });
-    let result = runner.run(&generate_messages, |initial_messages| {
-        let node_handles: Vec<_> = (1..=number_of_nodes)
+    let mut rng = StdRng::seed_from_u64(config.seed);
+
+    for test_number in 0..config.number_of_tests {
+        let entries = generator(&mut rng);
+
+        let node_handles: Vec<_> = (1..=config.number_of_nodes)
             .map(|i| (format!("n{}", i), spawn()))
             .collect();
 
-        let mut world = World::new(initial_messages, node_handles);
-        let trace = world.run_world();
-
-        match property(Trace(trace.to_vec())) {
-            Ok(()) => (),
-            Err(reason) => prop_assert!(false, "{}", reason),
-        }
-        Ok(())
-    });
-    match result {
-        Ok(_) => {
-            if persist_on_success {
-                persist_schedule_(Path::new("."), "success", trace_buffer)
-            }
+        let mut world = World::new(entries.clone(), node_handles);
+
+        match world.run_world() {
+            Err((reason, history)) => display_failure(
+                test_number,
+                config.seed,
+                entries,
+                History(history.to_vec()),
+                trace_buffer.clone(),
+                reason,
+            ),
+            Ok(history) => match property(History(history.to_vec())) {
+                Ok(()) => continue,
```

Comment:
this `continue` seems extraneous, but does it hint that the `Err(reason)` case should actually `break`?